### Лабораторная № 3. Кластеризация методом DBSCAN
Работу выполнил Жембровский Даниил. Задача – проведение кластеризации методом DBSCAN

11 утверджений для оценки:
- V1: Я потребляю леденцы, чтобы освежить дыхание
- V2: Я потребляю леденцы, чтобы чувствовать себя увереннее
- V3: Я потребляю леденцы, чтобы освежить полость рта
- V4: Я потребляю леденцы, чтобы избавиться от неприятного вкуса во рту
- V5: Я потребляю леденцы как заменитель других кондитерских изделий
- V6: Я потребляю леденцы, когда мне хочется чего-нибудь сладкого
- V7: Леденцы помогают мне сконцентрироваться
- V8: Я потребляю леденцы, чтобы отвлечься на несколько минут и подумать
- V9: Я потребляю леденцы для того, чтобы облегчить боль в горле
- V10: Я потребляю леденцы для того, чтобы избавиться от заложенности носа
- V11: Я потребляю леденцы для улучшения самочувствия

In [1]:
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
#matplotlib.style.use('ggplot') потом интереса ради активировать
%matplotlib inline
import os
import numpy as np
from sklearn.cluster import DBSCAN
from scipy.cluster.hierarchy import dendrogram, fcluster, linkage
from sklearn import metrics

In [3]:
os.chdir('C:/Users/danil/Desktop/Саморазвитие/ГПН/Lab_3')

In [2]:
df = pd.read_csv("C:/Users/danil/Desktop/Саморазвитие/ГПН/Lab_3/Lab_3_data.csv", sep = ';')
df

V1  V2  V3  V4  V5  V6  V7  V8  V9  V10  V11
0     4   5   5   5   3   3   3   3   3    3    3
1     5   4   5   5   3   3   3   3   3    3    3
2     5   5   4   5   3   3   3   3   3    3    3
3     5   4   5   5   3   3   3   3   3    3    3
4     4   5   5   5   3   3   3   3   3    3    3
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...
333   3   3   3   3   3   3   5   5   3    3    3
334   3   3   3   3   3   3   5   5   3    3    3
335   3   3   3   3   3   3   5   5   3    3    3
336   3   3   3   3   3   3   5   5   3    3    3
337   3   3   3   3   3   3   5   5   3    3    3

[338 rows x 11 columns]

In [4]:
df.describe()

V1          V2          V3          V4          V5          V6  \
count  338.000000  338.000000  338.000000  338.000000  338.000000  338.000000   
mean     3.544379    3.526627    3.517751    3.411243    3.207101    3.204142   
std      0.955748    0.956410    0.990150    0.992311    0.867314    0.869725   
min      2.000000    2.000000    2.000000    2.000000    2.000000    2.000000   
25%      3.000000    3.000000    3.000000    3.000000    3.000000    3.000000   
50%      3.000000    3.000000    3.000000    3.000000    3.000000    3.000000   
75%      5.000000    5.000000    5.000000    4.000000    3.000000    3.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

               V7          V8          V9         V10         V11  
count  338.000000  338.000000  338.000000  338.000000  338.000000  
mean     3.254438    3.289941    3.603550    3.550296    3.565089  
std      0.844107    0.821784    0.969682    0.927082    0.942042  
min      2.000000    2.000000    2.000000    2.000000    2.000000  
25%      3.000000    3.000000    3.000000    3.000000    3.000000  
50%      3.000000    3.000000    3.000000    3.000000    3.000000  
75%      3.000000    3.000000    5.000000    5.000000    5.000000  
max      5.000000    5.000000    5.000000    5.000000    5.000000

Оценки от 2 до 5 в стандартизации данных нет необходимости

In [37]:
dbscan = DBSCAN(eps = 0.5, min_samples = 5, metric = 'euclidean')
dbscan.fit(df)
dbscan.labels_    #массив номеров кластеров 

array([ 0,  1, -1,  1,  0,  1, -1,  1,  2, -1, -1, -1,  3, -1, -1,  3,  3,
        3,  3,  3,  3,  3, -1,  3,  3,  3,  3, -1, -1, -1,  2,  2,  0,  2,
        2,  2,  2,  2,  2,  2,  2,  2,  4,  4,  4,  4,  4,  4,  4,  4, -1,
        4, -1,  2,  2,  2,  2,  2,  2,  2, -1, -1, -1,  2,  2, -1, -1,  2,
        2,  2,  2, -1,  2,  2,  2,  2, -1,  2,  2,  2, -1,  2,  2,  2,  2,
        1,  2,  2,  2,  2,  2,  2,  2,  2,  2,  0, -1, -1,  0,  0,  0, -1,
       -1,  0,  0,  0,  2,  2,  2,  2, -1,  5, -1,  5,  5,  5,  6,  6,  5,
        5,  6,  5,  7, -1,  6,  7,  7,  5,  5,  5,  5,  5,  6,  5,  5,  5,
        5,  5,  6, -1,  5,  6,  5,  7,  6,  5, -1,  5,  6,  7,  5,  5,  5,
        5,  6,  5,  5,  5,  5,  5,  5,  6,  5,  5,  5,  5, -1,  5,  7,  5,
       -1,  5, -1,  6,  5,  5,  7, -1, -1,  8,  8,  8, -1, -1, -1, -1,  5,
        5, -1,  8, -1, -1, -1,  8,  8, -1, -1,  5,  5,  5,  5,  5,  7,  5,
       -1, -1,  6,  6,  5,  7, -1, -1, -1, -1, -1,  7, -1,  6,  9, 10, 11,
       10, 11,  9,  9,  9

In [38]:
a = np.unique(dbscan.labels_, return_counts = True)  # определяем уникальные значения, получается 2 массива: 
# 1 - номера кластеров, 2 - количество элементов в них
a

(array([-1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13],
       dtype=int64),
 array([60, 10,  5, 49, 12,  9, 49, 15, 10,  6, 37, 10,  5,  9, 52],
       dtype=int64))

In [39]:
np.asarray(a).T # превращаю переменную а, содержащую 2 массива в единый массив и транспонирую его
# получается избыточное количество кластеров

array([[-1, 60],
       [ 0, 10],
       [ 1,  5],
       [ 2, 49],
       [ 3, 12],
       [ 4,  9],
       [ 5, 49],
       [ 6, 15],
       [ 7, 10],
       [ 8,  6],
       [ 9, 37],
       [10, 10],
       [11,  5],
       [12,  9],
       [13, 52]], dtype=int64)

In [40]:
# увеличиваю радиус eps
dbscan_1 = DBSCAN(eps = 3, min_samples = 4, metric = 'euclidean')
dbscan_1.fit(df)
dbscan_1.labels_
np.asarray(np.unique(dbscan_1.labels_, return_counts = True)).T

# слишком высокие значения, поэтому получается 1 кластер

array([[  0, 338]], dtype=int64)

In [41]:
eps = np.arange(1, 3, 0.1) # массив от 1 до 3 с промежутком 0.1 (т.к. range работает только с int)
points = np.arange(2, 5, 1)
max_sil = 0 
for i in eps:
    for j in points:
        dbscan = DBSCAN(eps = float(i), min_samples = float(j), metric = 'euclidean') # так как числа небольшие, 
        # то можно не преобразовывать их из numpy.float64 в int
        dbscan.fit(df)
        if len(np.unique(dbscan.labels_)) != 1: # поставил ограничение, так как могут образовываться массивы с 1 кластером
            # а это вызывает ошибку при просчёте силуэта
            sil = metrics.silhouette_score(df, dbscan.labels_)
            if max_sil <= sil:   # поставил такое условие, чтобы значение переменной max_sil достигло максимума, а затем выбрать
                #параметры dbscan с максимальным значением силуэт
                max_sil = sil
                print(np.asarray(np.unique(dbscan.labels_, return_counts = True)))

[[ -1   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16
   17  18  19  20]
 [ 11 101   2   3   2  49   3  15  10   2   3   6   4   3   4  37  15   9
   52   3   2   2]]
[[ -1   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16
   17  18  19  20]
 [ 11 101   2   3   2  49   3  15  10   2   3   6   4   3   4  37  15   9
   52   3   2   2]]
[[ -1   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16
   17  18  19  20]
 [ 11 101   2   3   2  49   3  15  10   2   3   6   4   3   4  37  15   9
   52   3   2   2]]
[[ -1   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16
   17  18  19  20]
 [ 11 101   2   3   2  49   3  15  10   2   3   6   4   3   4  37  15   9
   52   3   2   2]]
[[ -1   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16
   17  18  19  20]
 [ 11 101   2   3   2  49   3  15  10   2   3   6   4   3   4  37  15   9
   52   3   2   2]]


#### Результаты неудовлетворительные. Получается 20 кластеров. Силуэт не репрезентативен, значение данного критерия высокое, однакополучается слишком много кластеров

In [42]:
for i in eps:
    for j in points:
        comb = []
        dbscan = DBSCAN(eps = i, min_samples = j, metric = 'euclidean')
        dbscan.fit(df)
        if 4 <= (len(np.unique(dbscan.labels_))) <= 6:  
            comb.append(round(i,1))
            comb.append(round(j, 1))
            print(np.asarray(np.unique(dbscan.labels_, return_counts = True)))
            print(comb)
# границы количества кластеров подобраны эмпирически: нижняя - количество кластеров при ICA (читерство во благо), 
# верхняя - предположение о возможном наличии 2 дополнительных группах уникумов, так или иначе отличающихся от остальных

[[  0   1   2   3   4   5]
 [110 104   4  61  57   2]]
[1.8, 2]
[[ -1   0   1   2   3   4]
 [  2 110 104   4  61  57]]
[1.8, 3]
[[ -1   0   1   2   3   4]
 [  2 110 104   4  61  57]]
[1.8, 4]
[[  0   1   2   3   4   5]
 [110 104   4  61  57   2]]
[1.9, 2]
[[ -1   0   1   2   3   4]
 [  2 110 104   4  61  57]]
[1.9, 3]
[[ -1   0   1   2   3   4]
 [  2 110 104   4  61  57]]
[1.9, 4]
[[  0   1   2   3]
 [110 104   4 120]]
[2.0, 2]
[[  0   1   2   3]
 [110 104   4 120]]
[2.0, 3]
[[  0   1   2   3]
 [110 104   4 120]]
[2.0, 4]
[[  0   1   2   3]
 [110 104   4 120]]
[2.1, 2]
[[  0   1   2   3]
 [110 104   4 120]]
[2.1, 3]
[[  0   1   2   3]
 [110 104   4 120]]
[2.1, 4]
[[  0   1   2   3]
 [110 104   4 120]]
[2.2, 2]
[[  0   1   2   3]
 [110 104   4 120]]
[2.2, 3]
[[  0   1   2   3]
 [110 104   4 120]]
[2.2, 4]
[[  0   1   2   3]
 [110 104   4 120]]
[2.3, 2]
[[  0   1   2   3]
 [110 104   4 120]]
[2.3, 3]
[[  0   1   2   3]
 [110 104   4 120]]
[2.3, 4]
[[  0   1   2   3]
 [110 104   4 120]]
[

Методом DBSCAN 4 кластера не получаются с адекватным количеством респондентов.
Буду рассматривать удачные кластеризации, где получилось 6 кластеров. Значения eps и min_samples возьму из списка comb

In [43]:
dbscan_2 = DBSCAN(eps = 1.8, min_samples = 2, metric = 'euclidean')
dbscan_2.fit(df)
dbscan_2.labels_
np.asarray(np.unique(dbscan_2.labels_, return_counts = True))

# читерство, но честное. Сравнивая с результатами, полученными ICA и Kmeans, это самый близкий по количеству респондентов 
# в кластерах, хотя добавляется 2 лишних кластера

array([[  0,   1,   2,   3,   4,   5],
       [110, 104,   4,  61,  57,   2]], dtype=int64)

In [44]:
# реализую иерархический кластерный анализ без построения дендрограммы
link = linkage(df, 'ward', 'euclidean')
link[:10]

array([[  0.,   4.,   0.,   2.],
       [  1.,   3.,   0.,   2.],
       [  2.,   6.,   0.,   2.],
       [  5., 339.,   0.,   3.],
       [ 32., 338.,   0.,   3.],
       [  7., 341.,   0.,   4.],
       [  8.,  30.,   0.,   2.],
       [ 12.,  15.,   0.,   2.],
       [ 16., 345.,   0.,   3.],
       [ 17., 346.,   0.,   4.]])

In [45]:
# строим таблицу сопряженности
cross = pd.DataFrame()
cross['ICA'] = fcluster(link, 4, criterion = 'maxclust')
cross['DBSCAN'] = dbscan_2.labels_

tab = pd.crosstab(cross['ICA'], cross['DBSCAN'])
tab

DBSCAN    0    1  2   3   4  5
ICA                           
1         0  104  4   0   0  0
2       110    0  0   0   0  0
3         0    0  0   0  57  2
4         0    0  0  61   0  0

In [46]:
df_2 = df.copy()
df_2['cluster'] = fcluster(link, 4, criterion = 'maxclust')
df_2.groupby(df_2['cluster']).mean()

V1        V2        V3        V4        V5        V6        V7  \
cluster                                                                         
1        2.824074  2.833333  2.768519  2.620370  2.611111  2.583333  2.750000   
2        4.845455  4.800000  4.836364  4.681818  3.036364  3.036364  3.000000   
3        3.000000  2.966102  2.966102  2.915254  2.915254  2.949153  4.915254   
4        3.000000  3.000000  3.000000  3.000000  4.852459  4.852459  3.000000   

               V8        V9       V10       V11  
cluster                                          
1        2.916667  4.953704  4.814815  4.861111  
2        2.990909  3.072727  3.045455  3.045455  
3        4.915254  3.000000  3.000000  3.000000  
4        2.918033  2.754098  2.754098  2.754098

In [47]:
df_3 = df.copy()
df_3['cluster'] = dbscan_2.labels_
df_3.groupby(df_3['cluster']).mean()

V1        V2        V3        V4        V5        V6        V7  \
cluster                                                                         
0        4.845455  4.800000  4.836364  4.681818  3.036364  3.036364  3.000000   
1        2.855769  2.788462  2.721154  2.567308  2.634615  2.605769  2.778846   
2        2.000000  4.000000  4.000000  4.000000  2.000000  2.000000  2.000000   
3        3.000000  3.000000  3.000000  3.000000  4.852459  4.852459  3.000000   
4        3.000000  3.000000  3.000000  2.947368  2.947368  2.947368  4.912281   
5        3.000000  2.000000  2.000000  2.000000  2.000000  3.000000  5.000000   

               V8        V9       V10       V11  
cluster                                          
0        2.990909  3.072727  3.045455  3.045455  
1        2.913462  4.951923  4.817308  4.865385  
2        3.000000  5.000000  4.750000  4.750000  
3        2.918033  2.754098  2.754098  2.754098  
4        4.912281  3.000000  3.000000  3.000000  
5        5.000000  3.000000  3.000000  3.000000

Кластеры по DBSCAN:
- 0. 110 респондентов - для уверенности, свеждести дыхания и избавления от неприятного запаха 
- 1. 104 респондентов - вылечить боль в горле, избавиться от заложенности 
- 2. 4 респондента - считаю, что их можно причислить к кластеру 1 или они хотят вылечиться и стать увереннее
- 3. 61 респондентов - используют как заменитель сладкого
- 4. 57 респондентов - для концентрации и чёткости мышления
- 5. 2 респондента - считаю, что их можно причислить к кластеру 4 